src/connection-loop.ts

```diff
@@ -0,0 +1,229 @@
+    // connection loop.
+    await Promise.resolve(0);
+    await this._currentResolver.promise;
+  }
+
+  async run(): Promise<void> {
+    const durations: Durations = [];
+    let delay = MIN_DELAY;
+    let lastConnectionErrored = false;
+    const delegate = this._delegate;
+    const {
+      debounceDelay = DEBOUNCE_DELAY,
+      maxConnections = MAX_CONNECTIONS,
+      watchdogTimer,
+    } = delegate;
+
+    console.log('Starting connection loop', delegate.printTime());
+
+    while (!delegate.closed()) {
+      console.log('Waiting for a push', delegate.printTime());
+
+      if (watchdogTimer) {
+        await Promise.race([
+          this._pendingResolver.promise,
+          sleep(watchdogTimer),
+        ]);
+      } else {
+        await this._pendingResolver.promise;
+      }
+
+      // This resolvers is used to make the individual calls to push have the
+      // correct "duration".
+      const currentResolver = resolver();
+      this._currentResolver = currentResolver;
+      // Don't let this rejected promise escape.
+      currentResolver.promise.catch(() => 0);
+
+      if (debounceDelay) {
+        await sleep(debounceDelay);
+      }
+
+      // This resolver is used to wait for incoming push calls.
+      this._pendingResolver = resolver();
+
+      if (lastConnectionErrored || this._pushCounter >= maxConnections) {
+        console.log(
+          'Too many pushes. Waiting until one finishes',
+          delegate.printTime(),
+        );
+        await this._waitUntilAvailableConnection();
+      }
+
+      if (lastConnectionErrored || this._pushCounter > 0) {
+        if (lastConnectionErrored) {
+          console.log(
+            'Last connection errored. Sleeping for',
+            delay,
+            'ms',
+            delegate.printTime(),
+          );
+        } else {
+          console.log(
+            'More than one outstanding connection (',
+            this._pushCounter,
+            '). Sleeping for',
+            delay,
+            'ms',
+            delegate.printTime(),
+          );
+        }
+
+        if (lastConnectionErrored) {
+          await sleep(delay);
+        } else {
+          const timeSinceLastSend = Date.now() - this._lastSendTime;
+
+          console.log(
+            'XXX',
+            delay,
+            timeSinceLastSend,
+            delay > timeSinceLastSend,
+            delay - timeSinceLastSend,
+          );
+
+          if (delay > timeSinceLastSend) {
+            await sleep(delay - timeSinceLastSend);
+          }
+        }
+      }
+
+      this._pushCounter++;
+      (async () => {
+        const start = Date.now();
+        let err: unknown;
+        try {
+          this._lastSendTime = start;
+          await delegate.invokeExecute();
+        } catch (e) {
+          err = e;
+        }
+        const stop = Date.now();
+        lastConnectionErrored = err !== undefined;
+        delay = computeDelayAndUpdateDurations(
```

Comment:
You could eliminate `delay` and make this function one piece of non-local state easier to understand by computing `delay` right before you sleep on it. There's no difference in cost since either way you compute `delay` once for each connection. Just a question of doing it before connecting or after.

Reply:
done